### PR TITLE
abc generics

### PR DESCRIPTION
**Pull request by nhannah**
# ABC Generics

## CLA

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [x] The code changes follow best practices
- [x] Code changes are tested (add some information if not applicable)

## Description of the pull request
In order to make integration of the typescript generics more consistent we have ABC'd their order in every instance to not create confusion.

### CODE REVIEW

File: ts-test/unit-test.ts [hunk 1]
```diff
@@ -23,6 +22,7 @@ import {
   UpdateChannelAPIResponse,
   PartialUserUpdate,
   PermissionObject,
+  ConnectAPIResponse,
```
Comment (virdesai) [inline ts-test/unit-test.ts:25]: this not need to be abc'd?

File: ts-test/unit-test.ts [hunk 2]
```diff
@@ -41,33 +41,41 @@ type AttachmentType = UnknownType;
 type EventType = UnknownType;
 type MessageType = UnknownType;
 type ReactionType = UnknownType;
+type CommandType = string & {};
```
Comment (virdesai) [inline ts-test/unit-test.ts:44]: should we abc this ordering too?

File: src/connection.ts [hunk 3]
```diff
@@ -59,21 +60,21 @@ type Constructor<
...
+  userID: Constructor<ChannelType, CommandType, UserType>['userID'];
+  user: Constructor<ChannelType, CommandType, UserType>['user'];
+  userAgent: Constructor<ChannelType, CommandType, UserType>['userAgent'];
+  authType: Constructor<ChannelType, CommandType, UserType>['authType'];
+  apiKey: Constructor<ChannelType, CommandType, UserType>['apiKey'];
+  tokenManager: Constructor<ChannelType, CommandType, UserType>['tokenManager'];
+  messageCallback: Constructor<ChannelType, CommandType, UserType>['messageCallback'];
+  recoverCallback: Constructor<ChannelType, CommandType, UserType>['recoverCallback'];
+  eventCallback: Constructor<ChannelType, CommandType, UserType>['eventCallback'];
+  logger: Constructor<ChannelType, CommandType, UserType>['logger'];
```
Comment (virdesai) [inline src/connection.ts:77]: these are in usefulness order?

File: ts-test/unit-test.ts (hunk 2 above)
Comment (nhannah) [reply ts-test/unit-test.ts:44]: Unit test is Vish's playground to double check types

File: src/client.ts [hunk 4]
```diff
@@ -97,53 +97,49 @@ function isString(x: unknown): x is string {
...
-  wsConnection: StableWSConnection<ChannelType, UserType, CommandType> | null;
-  wsPromise: Promise<void | ConnectionOpen<ChannelType, UserType, CommandType>> | null;
-  setUserPromise: Promise<void | ConnectionOpen<
-    ChannelType,
-    UserType,
-    CommandType
-  >> | null;
+  wsConnection: StableWSConnection<ChannelType, CommandType, UserType> | null;
+  wsPromise: ConnectAPIResponse<ChannelType, CommandType, UserType> | null;
+  setUserPromise: ConnectAPIResponse<ChannelType, CommandType, UserType> | null;
```
Comment (virdesai) [inline src/client.ts:133]: don't want to abc these?

File: src/client.ts [hunk 5]
```diff
@@ -1795,43 +1791,43 @@ export class StreamChat<
   /**
    * updateMessage - Update the given message
    *
-   * @param {MessageResponse<MessageType, AttachmentType, ChannelType, ReactionType, UserType, CommandType>} message object, id needs to be specified
```
Comment (virdesai) [inline src/client.ts:1798]: don't need the rest of these Types anymore?

Comment (nhannah) [reply src/client.ts:1798]: previous paste error in the comment they don't type check so we missed it before.

File: src/connection.ts (hunk 3 above)
Comment (nhannah) [reply src/connection.ts]: we were mostly abcing the generics where it effected use, did these for you though.

**Review by virdesai — APPROVED** [review]
LGTM

**Comment by vishalnarkhede** [comment]
LGTM. Can you please run a quick check against `stream-chat-react` with these changes?

**Comment by nhannah** [comment]
> LGTM. @nhannah Can you please run a quick check against `stream-chat-react` with these changes?

On it

**Comment by nhannah** [comment]
> > LGTM. @nhannah Can you please run a quick check against `stream-chat-react` with these changes?
> 
> On it

Once we merge and make a new release we can update this PR with the correct release and it will match up: https://github.com/GetStream/stream-chat-react/pull/521

**Review by vishalnarkhede — APPROVED** [review]

**Pull request merged**